### PR TITLE
GTG: Add rewire to work with istanbul for coverage

### PR DESCRIPTION
**Pull request by robert-chiniquy**
Add rewire to TestFile.runTests() so when coverage is running, the test module has the covered resources in lib/ injected into its scope.

### CODE REVIEW

**Comment by sam-falvo** [comment]
:icecream: +1

File: lib/common.js [hunk 1]
```diff
@@ -26,6 +27,10 @@ var gex = require('gex');
 var underscore = require('underscore');
 var log = require('logmagic').local('whiskey.common');
 
+var required_as = require('required-as');
+var rewire = require('rewire');
+var _ = require('underscore');
```
Comment (Kami) [inline lib/common.js]: Style - elsewhere we reference it as `underscore`.

File: lib/common.js [hunk 2]
```diff
@@ -352,7 +358,30 @@ TestFile.prototype.runTests = function(callback) {
     function(callback) {
       var errName;
       try {
-        exportedFunctions = require(testModule);
+        if (self._covered) {
+          var fullModuleName = testModule;
```
Comment (Kami) [inline lib/common.js]: I would add a short comment which explains what this if branch does.

File: lib/common.js [hunk 3]
```diff
@@ -26,6 +27,10 @@ var gex = require('gex');
 var underscore = require('underscore');
 var log = require('logmagic').local('whiskey.common');
 
+var required_as = require('required-as');
```
Comment (Kami) [inline lib/common.js]: `requiredAs` for style consistency

**Comment by Kami** [comment]
Awesome, no need to set `NODE_PATH` environment anymore, good stuff.

File: lib/common.js [hunk 4]
```diff
@@ -352,7 +358,30 @@ TestFile.prototype.runTests = function(callback) {
...
+            fullModuleName = fullModuleName + '.js';
+          }
+          var requirements = required_as(fullModuleName);
+
+          // for every requirement in lib/, rewrite to lib-cov/
+          var newRequirements = _.map(requirements, function(req) {
+            if (req.source.match(/node_modules/) || !req.source.match(/lib\//)) {
+              return undefined;
+            }
+            req.source = req.source.replace(/lib\//, 'lib-cov/');
```
Comment (Kami) [inline lib/common.js:378]: Hm, just thinking out loud about edge cases.
1. We already skip `node_modules` above so it should be fine
2. We only replace first occurrence of `lib/` so if someone does crazy thing like `require('lib/foo/lib/foo')` it should still work

**Comment by robert-chiniquy** [comment]
One problem I've realized with this is a case like:

`var method = require('package').method;`

Fixable and as long as it isn't used in a test file it's not an issue.

**Comment by Kami** [comment]
@robert-chiniquy You mean directly referencing a method in the require statement?

Is so, this is pretty common and we also do it in quite a lot of places.

File: lib/common.js (hunk 4 above)
Comment (robert-chiniquy) [inline lib/common.js:378]: Yup that's what I'm thinking.

**Comment by robert-chiniquy** [comment]
@Kami Yeah I mean that but _only_ in the test module. The method in question will not be counted towards coverage. This can happen all it wants in the tested code.

This can be fixed, just requires some more clever AST digging than is in there currently.

File: lib/common.js (hunk 1 above)
Comment (robert-chiniquy) [inline lib/common.js]: Fixed in 383ab5b.

File: lib/common.js (hunk 2 above)
Comment (robert-chiniquy) [inline lib/common.js]: Added in 383ab5b.

File: lib/common.js (hunk 3 above)
Comment (robert-chiniquy) [inline lib/common.js]: Fixed in 383ab5b.

**Comment by robert-chiniquy** [comment]
All comments addressed. I hope this `require('package').method` thing isn't a blocker. If it is, I'll fix it now— otherwise I'll do it as soon as I can.

**Comment by Kami** [comment]
@robert-chiniquy Can you please merge master into your branch, resolve conflicts and make sure it builds?

File: example/test-coverage.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ var sprintf = require('sprintf').sprintf;
 
 var cwd = process.cwd();
 
-exec(sprintf('NODE_PATH=lib-cov %s/bin/whiskey --tests %s/example/test-success-with-coverage.js --timeout 6000 --coverage --coverage-reporter cli', cwd, cwd),
```
Comment (Kami) [inline example/test-coverage.js:10]: I actually want to improve this test / add another test which also asserts on the actual coverage numbers.

I'll create a separate branch based on this one and add another test which verifies the coverage thresholds.

Comment (robert-chiniquy) [inline example/test-coverage.js:10]: coolcoolcool :icecream:

**Comment by robert-chiniquy** [comment]
Rebased on master and tests pass.

**Comment by Kami** [comment]
(cleaning up the old issues)

All the issues seems to have been addressed and the change looks good so I'm going to merge it. Thanks!

**Pull request merged**